File: ios/iOSBaseProject/Domain/UseCases/Auth/VerifyEmailUseCase.swift
//
//  VerifyEmailUseCase.swift
//  BaatCheet
//
//  Domain layer - Email Verification Use Case
//

import Foundation

/// Use case for verifying user's email.
/// Validates the code and delegates to the repository.
protocol VerifyEmailUseCaseProtocol {
    func execute(email: String, code: String) async -> AuthResult
}

final class VerifyEmailUseCase: VerifyEmailUseCaseProtocol {
    private let repository: AuthRepositoryProtocol
    
    init(repository: AuthRepositoryProtocol) {
        self.repository = repository
    }
    
    func execute(email: String, code: String) async -> AuthResult {
        // Validate inputs
        guard !email.isEmpty else {
            return .failure(.invalidEmail)
        }
        
        guard !code.isEmpty else {
            return .failure(.verificationFailed)
        }
        
        guard code.count == 6 else {
            return .failure(.verificationFailed)
        }
        
        guard code.allSatisfy({ $0.isNumber }) else {
            return .failure(.verificationFailed)
        }
        
        // Delegate to repository
        return await repository.verifyEmail(email: email, code: code)
    }
}

File: android/app/src/main/java/com/baatcheet/app/domain/usecase/auth/VerifyEmailUseCase.kt
package com.baatcheet.app.domain.usecase.auth

import com.baatcheet.app.domain.model.AuthResult
import com.baatcheet.app.domain.repository.AuthRepository
import javax.inject.Inject

/**
 * Verify Email Use Case
 * 
 * Handles the business logic for email verification.
 */
class VerifyEmailUseCase @Inject constructor(
    private val authRepository: AuthRepository
) {
    /**
     * Execute email verification with code
     */
    suspend operator fun invoke(email: String, code: String): AuthResult {
        // Validate inputs
        if (email.isBlank()) {
            return AuthResult.Failure(IllegalArgumentException("Email cannot be empty"))
        }
        if (code.isBlank()) {
            return AuthResult.Failure(IllegalArgumentException("Verification code cannot be empty"))
        }
        if (code.length != 6) {
            return AuthResult.Failure(IllegalArgumentException("Verification code must be 6 digits"))
        }
        if (!code.all { it.isDigit() }) {
            return AuthResult.Failure(IllegalArgumentException("Verification code must contain only digits"))
        }
        
        return authRepository.verifyEmail(email, code)
    }
}
